### ISSUE TITLE
Scala Incremental Compilation for Multi-Module projects broken in 7.x

### ISSUE THREAD

**Issue by justinb99** [gradle/gradle#20101]
Scala Incremental Compilation for Multi-Module projects is broken in 7.x (tested in 7.4) when it works in 6.9.2.

### Expected Behavior
Proper incremental compilation on a single source file change.

### Current Behavior
Full recompilation of a module for a single source file change in that module.

### Context
Working in 6.9.2, but does not work in 7.x (7.3.3, 7.4)

### Steps to Reproduce 
We have a large, proprietary codebase, where nested classes are being invalidated for dependent modules.  The repo https://github.com/justinb99/gradle7-scala-incr-compile-issue trivially reproduces the problem.

 In 7.4, a debug log shows that numerous nested class files from dependent modules become invalidated, so even changing one source file in a module triggers a full rebuild of that module.  Based on the observed logs, it looks identical to a previous issue in the Maven Scala Plugin, and it appeared the proper analysis information wasn't being carried between modules: https://github.com/davidB/scala-maven-plugin/issues/368

Below is an abbreviation of our full project build logs running Gradle 7.4
```
2022-03-05T13:27:37.579-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] Invalidating '/project/module1/build/classes/scala/main/module1/SafeThrowable.class' because module1.SafeThrowable$SafeThrowableWrapper is now provided by /project/module1/source/.../SafeThrowable$SafeThrowableWrapper.class**

# 70 other invalidations of nested class files

Initial source changes: 
	removed:Set()
	added: Set()
	modified: Set(/one/single/scala/file.scala)
Invalidated products: Set()
External API changes: API Changes: Set()
Modified binary dependencies: Set(/all/nested/class/files/previously/identified...)
Initial directly invalidated classes: Set(/one/single/scala/file.scala)

2022-03-05T13:27:37.792-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] Recompiling all sources: number of invalidated sources > 50.0% of all sources
```

For the referenced sample project, after a full compilation, make a change to Independent.scala (and IndependentTest.scala) and recompile/re-run tests (`./gradlew :module2:scalaTest --debug`).  From the Zinc logs, even though only Independent.scala changes, the StringUtil.class is invalidated: 
> Invalidating '/justinb99/gradle7-scala-incr-compile-issue/module1/build/classes/scala/main/justinb99/module1/StringUtil.class' because justinb99.module1.StringUtil$StringExtension is now provided by /justinb99/gradle7-scala-incr-compile-issue/module1/build/classes/scala/main/justinb99/module1/StringUtil$StringExtension.class

Below is a portion of the relevant build log from https://github.com/justinb99/gradle7-scala-incr-compile-issue during incremental compilation running Gradle 7.4:
```
2022-03-07T14:21:43.774-0600 [INFO] [org.gradle.api.internal.tasks.scala.ZincScalaCompiler] Compiling with Zinc Scala compiler.
2022-03-07T14:21:44.215-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompiler] Inputs(compilers: Compilers(scalac: Analyzing compiler (Scala 2.12.14), javaTools: sbt.internal.inc.javac.JavaTools$$anon$1@6ff3a6c4), options: CompileOptions(classpath: [Ljava.io.File;@6daa65bb, sources: [Ljava.io.File;@ada1ea5, classesDirectory: /justinb99/gradle7-scala-incr-compile-issue/module2/build/classes/scala/main, scalacOptions: [Ljava.lang.String;@209bfb9a, javacOptions: [Ljava.lang.String;@52acaccb, maxErrors: 100, sourcePositionMapper: xsbti.compile.CompileOptions$1@6c159c98, order: Mixed, temporaryClassesDirectory: Optional.empty), setup: Setup(perClasspathEntryLookup: org.gradle.api.internal.tasks.scala.ZincScalaCompiler$EntryLookup@36e08889, skip: false, cacheFile: /justinb99/gradle7-scala-incr-compile-issue/module2/build/tmp/scala/compilerAnalysis/compileScala.analysis, cache: sbt.internal.inc.FreshCompilerCache@49d8f2e4, incrementalCompilerOptions: IncOptions(transitiveStep: 5, recompileAllFraction: 0.5, relationsDebug: false, apiDebug: false, apiDiffContextSize: 5, apiDumpDirectory: Optional.empty, classfileManagerType: Optional[TransactionalManagerType(backupDirectory: /justinb99/gradle7-scala-incr-compile-issue/module2/build/tmp/scala/classfileBackup/compileScala.bak, logger: org.gradle.api.internal.tasks.scala.SbtLoggerAdapter@3e1467e0)], useCustomizedFileManager: false, recompileOnMacroDef: Optional[false], useOptimizedSealed: false, storeApis: true, enabled: true, extra: {}, logRecompileOnMacro: true, externalHooks: org.gradle.api.internal.tasks.scala.ZincScalaCompiler$LookupOnlyExternalHooks@c02046b, ignoredScalacOptions: [Ljava.lang.String;@24388dd, strictMode: false), reporter: sbt.internal.inc.LoggedReporter@5219c7a0, progress: Optional.empty, extra: [Lxsbti.T2;@4faf8164), previousResult: PreviousResult(analysis: Optional[Analysis: 2 Scala sources, 4 classes, 1 binary dependency], setup: Optional[MiniSetup(output: SingleOutput(/justinb99/gradle7-scala-incr-compile-issue/module2/build/classes/scala/main), options: MiniOptions(classpathHash: [Lxsbti.compile.FileHash;@6406d58c, scalacOptions: [Ljava.lang.String;@591c08fb, javacOptions: [Ljava.lang.String;@2b4b0e1b), compilerVersion: 2.12.14, order: Mixed, storeApis: true, extra: [Lxsbti.T2;@501094d2)]))
2022-03-07T14:21:44.215-0600 [INFO] [org.gradle.api.internal.tasks.scala.ZincScalaCompiler] Prepared Zinc Scala inputs: 0.441 secs
2022-03-07T14:21:44.603-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] Invalidating '/justinb99/gradle7-scala-incr-compile-issue/module1/build/classes/scala/main/justinb99/module1/StringUtil.class' because justinb99.module1.StringUtil$StringExtension is now provided by /justinb99/gradle7-scala-incr-compile-issue/module1/build/classes/scala/main/justinb99/module1/StringUtil$StringExtension.class
2022-03-07T14:21:44.610-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] 
Initial source changes: 
	removed:Set()
	added: Set()
	modified: Set(/justinb99/gradle7-scala-incr-compile-issue/module2/src/main/scala/justinb99/module2/Independent.scala)
Invalidated products: Set()
External API changes: API Changes: Set()
Modified binary dependencies: Set(/justinb99/gradle7-scala-incr-compile-issue/module1/build/classes/scala/main/justinb99/module1/StringUtil.class)
Initial directly invalidated classes: Set(justinb99.module2.Independent)

Sources indirectly invalidated by:
	product: Set()
	binary dep: Set(/justinb99/gradle7-scala-incr-compile-issue/module2/src/main/scala/justinb99/module2/Dependency.scala)
	external source: Set()
2022-03-07T14:21:44.610-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] All sources are invalidated.
2022-03-07T14:21:44.803-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] Created transactional ClassFileManager with tempDir = /justinb99/gradle7-scala-incr-compile-issue/module2/build/tmp/scala/classfileBackup/compileScala.bak
2022-03-07T14:21:44.809-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] Initial set of included nodes: justinb99.module2.Independent
2022-03-07T14:21:44.814-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] Recompiling all sources: number of invalidated sources > 50.0% of all sources
2022-03-07T14:21:44.816-0600 [DEBUG] [org.gradle.api.internal.tasks.scala.ZincScalaCompilerFactory] About to delete class files:
	/justinb99/gradle7-scala-incr-compile-issue/module2/build/classes/scala/main/justinb99/module2/Dependency.class
	/justinb99/gradle7-scala-incr-compile-issue/module2/build/classes/scala/main/justinb99/module2/Dependency$.class
	/justinb99/gradle7-scala-incr-compile-issue/module2/build/classes/scala/main/justinb99/module2/Independent.class
	/justinb99/gradle7-scala-incr-compile-issue/module2/build/classes/scala/main/justinb99/module2/Independent$.class
```

### Your Environment
macOS 12.2.1, Gradle 7.4, openjdk 11.0.2, Scala 2.12.14

**Comment by mernst-github** [comment]
Running into the same issue on 7.5.1, this is quite a productivity killer. https://xkcd.com/303/

**Comment by mernst-github** [comment]
I did some digging/logging around the invalidation code and I'm pretty sure something's wrong here:
https://github.com/gradle/gradle/blob/master/subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java#L232

I see two `Stamp`s of different types compared, where `previousAnalysis` holds stamps of type `farm` whereas gradle uses `lastModified`. So they simply cannot compare equal. I guess that's something that changed in zinc and gradle needs to follow suit?

**Comment by ljacomet** [comment]
@tgodzik @dubinsky @ijuma Hey folks, any chance someone with Zinc knowledge can chime in?
Is there some way to resolve this without having to revamp the way Gradle works?

**Comment by mernst-github** [comment]
Some more investigation: if the above is fixed on the Gradle side (use hash instead of last modified), the example starts working. zinc can now properly recognize that moduleA.jar has not changed. But this is still not sufficient unfortunately.

_In combination with the java-library plugin_, the inter-module dependencies turn from "moduleB class" -> "moduleA jar" into more fine-grained "moduleB class -> moduleA classfile" dependencies (as [advertised](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_classes_usage), and the way one wants it to be.

However, zinc is not really prepared for this - if classpath entries are directories its own comparison does bad things (compares classfiles to classpath directories, see a comment I left on [zinc](https://github.com/sbt/zinc/pull/712/files#r1002970496) . When I hack zinc to properly compare classfile to classfile it kind of starts working, but even then there seem to be bugs regarding dependency tracking with inner classes.

Somewhat surprising.

**Comment by tgodzik** [comment]
@ljacomet I haven't noticed the different hash types when upgrading zinc, sorry about that! I see that @mernst-github has already done a great job investigating and we should probably update the hash used for changedBinaries. Would you be able to make that change in Gradle @mernst-github? Or do you need help with that?

As for the second point, I am not sure I understand what is going on, but I see that Eugene has already turned it into an issue:
https://github.com/sbt/zinc/issues/1140

I will monitor that issue and see if it's something I can help with.

**Comment by ljacomet** [comment]
@tgodzik Thanks for responding!

@mernst-github Thanks for the analysis!

We could balance the jar to classfile issue by having Scala request jars and not classes. While I agree it is conceptually better, losing incremental compilation for saving on creating a JAR does not seem like a good trade-of.
The Groovy plugin already does something like that since it may need files in resources as well:

https://github.com/gradle/gradle/blob/3193f06be6e89337b2b1c055708ea094ecec60d4/subprojects/plugins/src/main/java/org/gradle/api/plugins/GroovyBasePlugin.java#L137-L142

Although for Scala, I would request `LibraryElements.JAR` since that's what Zinc requires for now.

**Comment by mernst-github** [comment]
NOT A CONTRIBUTION

There's no need for the Gradle plugin to even have an [external lookup implementation](https://github.com/gradle/gradle/blob/master/subprojects/scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompiler.java#L214-L287) - [zinc's default](https://github.com/sbt/zinc/blob/develop/internal/zinc-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala#L407-L418) is perfectly adequate and correct for the case of using _just_ the Scala plugin (since that creates jar->jar dependencies).

The classes dependencies you get when you add the java-library plugin, and the ensuing "classpath entry is a directory" is a separate issue that can be fixed in zinc. I wouldn't hobble the plugin because of that, especially when the zinc version is configurable.

**Comment by mernst-github** [comment]
Sorry for the wait while I got the necessary things together. I'm proposing to completely remove Gradle's faulty "has binary dependency changed" customization in the above PR, and rely on Zinc's default. This is working well for us (modulo https://github.com/sbt/zinc/issues/1140).

**Comment by ljacomet** [comment]
* Fixed through #22786
